chore: DLT-1513 remove multiple exports

juliodialpad opened the pull request:
# Remove multiple exports

## Obligatory GIF (super important!)

![Obligatory GIF](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExbnVmenF4Y2RhMHl1MmV5YTB5amowbHlwM2lmbjZ5ZGZkbmFrNmtsdCZlcD12MV9naWZzX3NlYXJjaCZjdD1n/paLMiK81cGM2KPdtaT/giphy.gif)

## :hammer_and_wrench: Type Of Change

These types will not increment the version number, but will still deploy to documentation site on release:

- [x] Build system

## :book: Jira Ticket

https://dialpad.atlassian.net/browse/DLT-1513?atlOrigin=eyJpIjoiZjY0NmMzYjE1Yzc4NDdhMGE5YmM4MTVlNmNiYWEwMzEiLCJwIjoiaiJ9

## :book: Description

- Removed exports from dialtone-vue2, dialtone-vue3 and dialtone
- Moved dialtone.css file from `lib/dist/css` to `lib/dist` to solve issues with fonts not being found on webpack + vue 2.6
- Removed `{ reactive }` import from DtHovercard as it is vue 2.7+ only and it will not work on vue 2.6
- Greatly reduced the bundle size by doing the following:
  - Created a new package dialtone-emojis to externalize this dependency
  - Externalized all dependencies

## :bulb: Context

We were starting to have a lot of exports on package.json which would be hard to maintain in the future, we needed to go back to having just two exports (main and style)

## :pencil: Checklist

For all PRs:

- [x] I have ensured no private Dialpad links or info are in the code or pull request description (Dialtone is a public repo!).
- [x] I have reviewed my changes.
- [x] I have added all relevant documentation.
- [x] I have considered the performance impact of my change.

For all Vue changes:

- [x] I have added / updated unit tests.
- [x] I have made my changes in Vue 2 and Vue 3. Note: you may sync your changes from Vue 2 to Vue 3 (or vice versa) using the `./scripts/dialtone-vue-sync.sh` script.

If new component:

- I am exporting any new components or constants:
  - [x] from the index.js in the root (either `packages/dialtone-vue2` or `packages/dialtone-vue3`).

## :crystal_ball: Next Steps

- Add tree shake, this will be done on a separate PR due too many files changed already on this.

## :camera: Screenshots / GIFs
- Before
![Screenshot 2024-03-20 at 12 07 52 a m](https://github.com/dialpad/dialtone/assets/87546543/615e82cb-5fb4-4529-a343-14fb1ccde2f6)
![image](https://github.com/dialpad/dialtone/assets/87546543/84056a1f-22e3-46df-bacf-64384af66f56)


- After
![Screenshot 2024-03-20 at 12 05 06 a m](https://github.com/dialpad/dialtone/assets/87546543/30ead525-2db3-42b4-b202-a889d0a57379)
![Screenshot 2024-03-20 at 12 09 50 a m](https://github.com/dialpad/dialtone/assets/87546543/e11b209c-d920-48e5-b1c5-238c4cdf899a)


## :link: Sources

- https://medium.com/@rjayakumarreddy887/create-a-ui-component-library-that-is-tree-shakable-cffbbc4d1829
- https://dev.to/rafaelogic/optimizing-vuejs-performance-a-guide-to-tree-shaking-with-webpack-and-vite-3if7
- https://blog.theodo.com/2021/04/library-tree-shaking/
- https://github.com/vitejs/vite/discussions/14676
- https://dev.to/receter/how-to-create-a-react-component-library-using-vites-library-mode-4lma

juliodialpad committed:
add dialtone-emojis package

juliodialpad committed:
add dialtone-emojis to dialtone-vue2

juliodialpad committed:
remove multiple exports and externalize packages

juliodialpad committed:
fix css imports

juliodialpad committed:
add peerDeps

juliodialpad committed:
remove reactive

juliodialpad committed:
remove unneeded dependency

juliodialpad committed:
sort exports

juliodialpad committed:
copy changes to vue 3

juliodialpad committed:
package improvements

juliodialpad committed:
package improvements

juliodialpad committed:
fix issues

juliodialpad committed:
Merge branch 'staging' into remove-multiple-exports

juliodialpad committed:
optimize external deps

juliodialpad committed:
remove unneeded files

juliodialpad committed:
fix test issues

juliodialpad committed:
fix issues

juliodialpad committed:
update docs

juliodialpad committed:
update docs

juliodialpad committed:
Merge branch 'staging' into remove-multiple-exports

# Conflicts:
#	packages/stylelint-plugin-dialtone/package.json

iropolo commented:
Seems like a huge improvement, do this impact how to use the library that we should notify or all this is just internal refactor?

juliodialpad commented:
> Seems like a huge improvement, do this impact how to use the library that we should notify or all this is just internal refactor?

I removed some exports: dialtone-vue/emoji, dialtone-vue/message_input and dialtone-vue/directives. Updated docs accordingly but a migration is going to be required.

Some of those exports might come back with the tree-shaking improvement I'm working on currently, I think I'll create a branch off of this one and finish the tree shaking improvement before we can merge this one.

braddialpad commented:
> Some of those exports might come back with the tree-shaking improvement I'm working on currently, I think I'll create a branch off of this one and finish the tree shaking improvement before we can merge this one.

Yeah I agree we should do this.

juliodialpad committed:
add formats to dialtone-vue

juliodialpad committed:
Merge branch 'staging' into remove-multiple-exports

braddialpad commented on packages/dialtone-vue2/vite.config.js:
```diff
@@ -10,21 +10,53 @@ const __dirname = path.dirname(__filename);
+      formats: ['es', 'cjs'],
     },
     rollupOptions: {
       external: [
+        /@dialpad/,
+        /@linusborg/,
+        /@tiptap/,
+        /date-fns/,
+        /emoji-regex/,
+        /emoji-toolkit/,
+        /tippy\.js/,
+        /prosemirror/,
         'vue',
```
If we externalize all these dependencies doesn't that mean the consumer will have to manually install these in order to run the library?

juliodialpad replied:
Those are already listed under `dependencies` so npm/pnpm/yarn will install `dependencies` and `peerDependencies`

braddialpad replied:
Cool, didn't actually realize it worked like this. I guess that makes sense as vite is totally separate from npm so you have to tell it what to bundle

braddialpad replied:
Could also use a vite plugin so we don't have to update this manually: https://www.npmjs.com/package/vite-plugin-externalize-deps

juliodialpad replied:
Tried that plugin and didn't worked well, maybe I was doing something wrong, I'll give it another shot.

Also tried importing dependencies and peerDependencies from package.json but needed to use `assert { type: 'json' }` which is too experimental to rely on it.

braddialpad replied:
Yeah it's fine to do manually for now. not that big of deal

braddialpad reviewed (approved):
Looks good, wait on merging for other PR as mentioned.

juliodialpad committed:
perf: dialtone-icons tree shaking (#226)

juliodialpad committed:
Merge remote-tracking branch 'origin/staging' into remove-multiple-exports

# Conflicts:
#	apps/dialtone-documentation/docs/about/whats-new/posts/2024-3-20.md
#	packages/dialtone-vue2/recipes/conversation_view/message_input/message_input.vue
#	packages/dialtone-vue3/recipes/conversation_view/message_input/message_input.vue

juliodialpad committed:
perf: DLT-1652 dialtone-vue tree shaking (#238)

juliodialpad committed:
fix alignment issues

juliodialpad committed:
Merge branch 'staging' into remove-multiple-exports

juliodialpad committed:
update mono-package dist files

juliodialpad committed:
update lock file

juliodialpad committed:
fix alignment issues

juliodialpad committed:
Merge branch 'staging' into remove-multiple-exports

# Conflicts:
#	packages/dialtone-vue2/package.json
#	packages/dialtone-vue3/package.json

juliodialpad committed:
fix alignment issues

juliodialpad committed:
remove icon vertical align

juliodialpad committed:
add show-skeleton prop

juliodialpad committed:
fix issues

juliodialpad committed:
vue 3 version

juliodialpad committed:
chore: NO-JIRA update mono package exports (#245)

juliodialpad committed:
update pnpm lock

juliodialpad committed:
update percy

juliodialpad committed:
remove includes

juliodialpad committed:
update percy version

juliodialpad committed:
update exports

juliodialpad committed:
change percy config

juliodialpad committed:
update lock file

juliodialpad commented:
Been having some issues with unstable visual tests due to lazy-loading icons, will investigate more about this later if they continue failing.
